src/main/scala/Ring.scala

```diff
@@ -0,0 +1,185 @@
+    randomize: Boolean = false)
+    (implicit p: Parameters) extends LazyModule {
+
+  val node = TLNexusNode(
+    clientFn  = { seq =>
+      seq(0).copy(
+        minLatency = seq.map(_.minLatency).min,
+        clients = (TLXbar.mapInputIds(seq) zip seq) flatMap { case (range, port) =>
+          port.clients map { client => client.copy(
+            sourceId = client.sourceId.shift(range.start)
+          )}
+        })
+    },
+    managerFn = { seq =>
+      val fifoIdFactory = TLXbar.relabeler()
+      seq(0).copy(
+        minLatency = seq.map(_.minLatency).min,
+        endSinkId = TLXbar.mapOutputIds(seq).map(_.end).max,
+        managers = seq.flatMap { port =>
+          require (port.beatBytes == seq(0).beatBytes,
+            s"Ring data widths don't match: ${port.managers.map(_.name)} has ${port.beatBytes}B vs ${seq(0).managers.map(_.name)} has ${seq(0).beatBytes}B")
+          val fifoIdMapper = fifoIdFactory()
+          port.managers map { manager => manager.copy(
+            fifoId = manager.fifoId.map(fifoIdMapper(_))
+          )}
+        })
+    })
+
+  lazy val module = new LazyModuleImp(this) with HasTLNetwork {
+    val (io_in, edgesIn) = node.in.unzip
+    val (io_out, edgesOut) = node.out.unzip
+    val nIn = edgesIn.size
+    val nOut = edgesOut.size
+    val nNodes = nIn + nOut
+    val idBits = log2Ceil(nNodes)
+
+    val forwardIds = (0 until nOut).map(id => (id + nIn).U(idBits.W))
+    val backwardIds = (0 until nIn).map(_.U(idBits.W))
+    val networkName = "TLRingNetwork"
+
+    if (nIn > 1 || nOut > 1) {
+      val aRing = Module(new NetworkRing(
+        nIn, nOut, new TLBundleA(commonBundle),
+        buffer.a, true, randomize))
+
+      val bRing = Module(new NetworkRing(
+        nOut, nIn, new TLBundleB(commonBundle),
+        buffer.b, false, randomize))
+
+      val cRing = Module(new NetworkRing(
+        nIn, nOut, new TLBundleC(commonBundle),
+        buffer.c, true, randomize))
+
+      val dRing = Module(new NetworkRing(
+        nOut, nIn, new TLBundleD(commonBundle),
+        buffer.d, false, randomize))
+
+      val eRing = Module(new NetworkRing(
+        nIn, nOut, new TLBundleE(commonBundle),
+        buffer.e, true, randomize))
+
+      io_in.zipWithIndex.foreach { case (in, i) =>
+        connectInput(i, in,
+          aRing.io.in(i),
+          bRing.io.out(i),
+          cRing.io.in(i),
+          dRing.io.out(i),
+          eRing.io.in(i))
+      }
+
+      io_out.zipWithIndex.foreach { case (out, i) =>
+        connectOutput(i, out,
+          aRing.io.out(i),
+          bRing.io.in(i),
+          cRing.io.out(i),
+          dRing.io.in(i),
+          eRing.io.out(i))
+      }
+    } else {
+      io_out.head <> io_in.head
+    }
+  }
+}
+
+class RingSystemBus(params: SystemBusParams, buffer: TLNetworkBufferParams)
+    (implicit p: Parameters) extends SystemBus(params) {
+  private val system_bus_ring = LazyModule(new TLRingNetwork(buffer))
```

Comment:
If this is private can Chisel still name it sanely?

Reply:
Yes it does. I checked the generated verilog and it has the right name.

Reply:
I guess I should have asked this before, but is there a particular reason it needs to be private? Correct me if I'm wrong, but this would prevent us from accessing this via an aspect wouldn't it? I could see us wanting to do that with the floorplanning framework.

Reply:
The equivalent structure in the standard SystemBus was private, and I thought there was some reason for that, but apparently not.